src/Initializer/BatchRecorders/DynamicRuptureRecorder.cpp

```diff
@@ -0,0 +1,104 @@
+#include "Recorders.h"
+#include <Kernels/Interface.hpp>
+#include <yateto.h>
+
+using namespace device;
+using namespace seissol::initializers;
+using namespace seissol::initializers::recording;
+
+
+void DynamicRuptureRecorder::record(DynamicRupture &handler, Layer &layer) {
+  setUpContext(handler, layer);
+  recordDofsTimeEvaluation();
+  recordSpaceInterpolation();
+}
+
+
+void DynamicRuptureRecorder::recordDofsTimeEvaluation() {
+  real** timeDerivativePlus = currentLayer->var(currentHandler->timeDerivativePlus);
+  real** timeDerivativeMinus = currentLayer->var(currentHandler->timeDerivativeMinus);
+  real* idofsPlus = static_cast<real *>(currentLayer->getScratchpadMemory(currentHandler->idofsPlusOnDevice));
+  real* idofsMinus = static_cast<real *>(currentLayer->getScratchpadMemory(currentHandler->idofsMinusOnDevice));
+
+  if (currentLayer->getNumberOfCells()) {
+    std::vector<real *> timeDerivativePlusPtrs{};
+    std::vector<real *> timeDerivativeMinusPtrs{};
+    std::vector<real *> idofsPlusPtrs{};
+    std::vector<real *> idofsMinusPtrs{};
+
+    const size_t idofsSize = tensor::Q::size();
+    for (unsigned face = 0; face < currentLayer->getNumberOfCells(); ++face) {
+      timeDerivativePlusPtrs.push_back(timeDerivativePlus[face]);
+      timeDerivativeMinusPtrs.push_back(timeDerivativeMinus[face]);
+      idofsPlusPtrs.push_back(&idofsPlus[face * idofsSize]);
+      idofsMinusPtrs.push_back(&idofsMinus[face * idofsSize]);
+    }
+
+    if (!idofsPlusPtrs.empty()) {
+      ConditionalKey key(*KernelNames::DrTime);
+      checkKey(key);
+
+      (*currentTable)[key].content[*EntityId::DrDerivativesPlus] = new BatchPointers(timeDerivativePlusPtrs);
+      (*currentTable)[key].content[*EntityId::DrDerivativesMinus] = new BatchPointers(timeDerivativeMinusPtrs);
+      (*currentTable)[key].content[*EntityId::DrIdofsPlus] = new BatchPointers(idofsPlusPtrs);
+      (*currentTable)[key].content[*EntityId::DrIdofsMinus] = new BatchPointers(idofsMinusPtrs);
```

Comment:
So, `currentTable` is a member of `LTStree`. Each layer has its own conditional table 
```
  // a part of class AbstractRecorder
  void setUpContext(LtsT &handler, Layer &layer) {
    currentTable = &(layer.getCondBatchTable());
```
Here is a definition:
```
using ConditionalBatchTableT =
    std::unordered_map<ConditionalKey, BatchTable, ConditionalHash<ConditionalKey>>;
}
```

In its turn, BatchTable is a wrapper of std::array which is responsible for deleting `BatchPointers`
```
struct BatchTable {
public:
  BatchTable() {
    for (auto &item : content) {
      item = nullptr;
    }
  }
  ~BatchTable() {
    for (auto &item : content) {
      delete item;
    }
  }

  std::array<BatchPointers *, *EntityId::Count> content{};
};
```

Here is a constructor and destructor of BatchPointers:
```
  explicit BatchPointers(std::vector<real *> collectedPointers)
      : pointers(std::move(collectedPointers)), devicePtrs(nullptr) {
    if (!pointers.empty()) {
      devicePtrs = (real **)device.api->allocGlobMem(pointers.size() * sizeof(real *));
      device.api->copyTo(devicePtrs, pointers.data(), pointers.size() * sizeof(real *));
    }
  }

  virtual ~BatchPointers() {
    if (devicePtrs != nullptr) {
      device.api->freeMem(devicePtrs);
      devicePtrs = nullptr;
    }
  }
```
LTSTree belongs to MemoryManager and it belongs to SeisSol class. Currently MemoryManager is a unique pointer in SeisSol class
```
  std::unique_ptr<initializers::MemoryManager> m_memoryManager{nullptr};
```

There is a method which we call from fortran at the very of an execution
```
void deleteMemoryManager() {
  m_memoryManager.reset(nullptr);
}
```
which is supposed to call m_ memoryManager destructor.

So, in short, according to me, everything should be deleted

Reply:
yes, I record raw pointers to GPU memory which will later be passed to yateto/gemmforge generated GPU kernels

Reply:
Thanks for the very detailed explanation! Are you sure that the destructors of the BatchPointers is actually called? I don't think so, unless I'm missing something. Could you maybe check? This is not an huge issue but it makes debugging with e.g. valgrind very hard because you get some strange output at the end of the program.

Reply:
I didn’t use valgrid because there is always some clutter caused by MPI. I know that there are some filter files for OpenMPI but it didn’t work for me. Tell me if you use valgrind for SeisSol.

When an LTS layer dies and destroys its instance of ConditionalBatchTableT (which is unordered_map) using its destructor. Each STL container calls each of the contained element's destructors while destructing itself. Therefore, destructor of BatchTable instances are going to be called which, in turn, are going to delete all BatchPointers.

Reply:
I added a print statement to the BatchPointers destructor to just as a sanity check and ran SeisSol. At the end I got a bunch of prints on my console. So, it wokrs

Reply:
Ah, sorry. I misunderstood your explanation. Read it again, and it's clear.